First implementation of MvccDeletePlugin

julianmenzler opened the pull request:
Let's review our changes as a whole.

d-justen committed:
add experimental update benchmark

engelfa committed:
Add MVCC Benchmark.

julianmenzler committed:
add benchmark approach (MVCC Updates)

julianmenzler committed:
Work on MVCC_Benchmark_Fiixture class

	- structure benchmark fixture
	- create some ToDos

d-justen committed:
remove update_benchmark, update CMakeLists

d-justen committed:
Add MvccDeleteTest

d-justen committed:
Make tests work

d-justen committed:
Add dummy mvcc delete plugin

d-justen committed:
Fix loading issue

d-justen committed:
Clean-up plugin template

d-justen committed:
Add test for chunk with partial invalid rows

julianmenzler committed:
implemented MVCC benchmark with fixture class

julianmenzler committed:
fix bug, rename mvcc benchmark filenames, implement local mvcc-benchmark-table-template

engelfa committed:
Create test.

julianmenzler committed:
add logs & add some missing operator->execute() calls & adjust formatting

julianmenzler committed:
Fixed bug using the Google Benchmark suite & change benchmark logic

- Benchmark was called without an argument, which lead to problems in the fixture class.
  This was due to the use of the wrong macro:
  → BENCHMARK_F instead of BENCHMARK_DEFINE_F

- Changed benchmark logic. Within the benchmark routine UPDATE is no longer called.
  → Instead, just a VALIDATE operator is used, which does not alter the table anymore.
    → This way benchmark preconditions are preserved throughout multiple iterations

julianmenzler committed:
Adjust chunk size and argument range for benchmark & add UPDATE benchmark

engelfa committed:
Fix TransactionContext test.

julianmenzler committed:
Add markdown-file with regex for benchmark visualization with python

d-justen committed:
Implement invalid row per chunk counting

d-justen committed:
Implement experimental mvcc delete

julianmenzler committed:
Add class MvccManager & add draft implementation for logical delete

julianmenzler committed:
Add assertion for edge case in deleteChunkLogically()

julianmenzler committed:
Rename function names and class MvccManager

	- function names adjusted to match project code style
	- MvccManager →  to MvccDeleteManager

d-justen committed:
Fix invalid row bug

julianmenzler committed:
Add MvccDeleteManagerTest

julianmenzler committed:
Implement forgotten commit() in MvccDeleteManager

d-justen committed:
remove couts, fix increment input parameter

d-justen committed:
Add cleanup id functionality

julianmenzler committed:
Adjust comments

d-justen committed:
move cleanup check into get table operator

julianmenzler committed:
Use UPDATE operator for logical delete of chunks

  - Before INSERT and DELETE were used. UPDATE is more elegant and should not introduce a noticeable overhead.

julianmenzler committed:
Rename cleanup_id to cleanup_commit_id in Chunk class

julianmenzler committed:
Add DebugAssert & rename setter and getter of cleanup_commit_id

julianmenzler committed:
Cherry-pick changes of Chunk class

  - cleanup_commit_id
  - invalid_row_count

julianmenzler committed:
Change logical delete implementation to set chunk cleanup_commit_id and return success value

d-justen committed:
Implement cleanup commit id check in get_table

d-justen committed:
Format code

d-justen committed:
Resolve merge conflicts

julianmenzler committed:
Add logging for debug purposes & use namespace opossum in MvccDeleteManager

julianmenzler committed:
Remove logging

julianmenzler committed:
Change implementation to use the GetTable-operator with excluded chunks instead of a temporary table

julianmenzler committed:
Fix logical delete by adding and using a _get_referencing_table() method

Previous implementations did not work since they conflicted with the implementation of the Delete resp. Update operator:
  - Delete operator requirement: "values_to_delete must (...) reference the table specified by table_name."
  - The GetTable operator in conjunction with set_excluded_chunk_ids() creates a new temporary table, which is not stored in the StorageManager. Following operators like Validate create tables referencing the temporary table, but not the original table.

_get_referencing_table() creates a temporary table, just like GetTable with set_excluded_chunk_ids(). But in comparison it references to the original table, which makes the difference when using the Update resp. Delete operators.

julianmenzler committed:
Extend test case for logical delete

  - Add utility method _get_value_from_table
  - Do full table structure check after each modification

julianmenzler committed:
Merge branch 'mvcc_delete_plugin' into mvcc-logical-delete & reformat

# Conflicts:
#	src/benchmark/CMakeLists.txt
#	src/benchmark/operators/mvcc_benchmark.cpp
#	src/lib/storage/chunk.cpp

julianmenzler committed:
Merge https://github.com/hyrise/hyrise into develop

julianmenzler committed:
Merge branch 'mvcc-logical-delete' into develop

engelfa committed:
Add python benchmark visualizer.

julianmenzler committed:
Change paths

d-justen committed:
Fix load plugin

julianmenzler committed:
Add draft implementation for physcal delete & improve Logical Delete test

d-justen committed:
move logical deletion into plugin

julianmenzler committed:
Implement expired-status in TransactionContext class

julianmenzler committed:
Implement container for "active-snapshot-ids" in TransactionManager

julianmenzler committed:
Implement snapshot-commit-id-check in Physical Delete

julianmenzler committed:
Fix unused variable

julianmenzler committed:
Fix DebugAssert & change expired-transaction-definition

	- transactions that did not finish their rollback might become a problem for the Physical Delete

julianmenzler committed:
Change interface and class name of MvccDeleteManager & implement physical_delete_queue logic in MvccDeletePlugin class

julianmenzler committed:
Add tests for physical delete & fix tests

julianmenzler committed:
Add logging for debugging

	- Issue in TransactionManager: Unexpected snapshot-commit-ids in _active_snapshot_commit_ids

julianmenzler committed:
Add file with Timer struct for benchmarking purposes

julianmenzler committed:
Fix bug in TransactionManager and remove related logging

Reset of _active_snapshot_commit_ids was missing.
Therefore, snapshot-commit-ids in the multiset were carried over from test to test.
This prevented the physical delete implementation to pass the check (cleanup_commit_id < lowest_snapshot_commit_id)

julianmenzler committed:
Fix segmentation fault error

julianmenzler committed:
Remove expired-state in TransactionContext and use the destructor solely to notify the TransactionManager about finish

julianmenzler committed:
Fix physical delete & finish two test cases for it

julianmenzler committed:
Structure comments in MvccDeleteManager tests & format some other files

julianmenzler committed:
Merge branch 'mvcc-physical-delete' into develop

julianmenzler committed:
Merge branch 'mvcc-physical-delete' into mvcc_delete_plugin

# Conflicts:
#	src/plugins/mvcc_delete_plugin.cpp
#	src/plugins/mvcc_delete_plugin.hpp

julianmenzler committed:
Merge classes MvccDelete/MvccDeletePlugin + rename corresponding test-classes

	- Integrate:
		mvcc_delete_manager.hpp/cpp → mvcc_delete_plugin.hpp/cpp
		implement FRIEND_TEST macro to access private/protected members from Google test classes
	- Rename:
		mvcc_delete_test.cpp → mvcc_delete_plugin_test.cpp
		mvcc_delete_manager_test.cpp → mvcc_delete_test.cpp
	- Adjust CMakeLists accordingly

julianmenzler committed:
Fix Linker error

julianmenzler committed:
Fix compile error, which was caused by a missing constructor in ChunkSpecifier struct

d-justen committed:
Create dummy mvcc delete class

d-justen committed:
Implement simple console output benchmark

d-justen committed:
Edit console output

d-justen committed:
Merge update fix into develop

d-justen committed:
merge

d-justen committed:
Merge branch 'develop' of github.com:d-justen/hyrise into mvcc_benchmark

julianmenzler committed:
Fix bug in TransactionManager

d-justen committed:
Add invalidated row amount method

d-justen committed:
Add file io

julianmenzler committed:
Implement functions begin_cleanup() & end_cleanup()

d-justen committed:
resolve merge conflicts

d-justen committed:
Merge branch 'develop' of github.com:d-justen/hyrise into mvcc_benchmark

julianmenzler committed:
Implement locking for _physical_delete_queue

julianmenzler committed:
Implement basic threading for running MvccDeletePlugin asynchronously

d-justen committed:
Refactor benchmark

d-justen committed:
Merge branch 'mvcc_delete_plugin' of github.com:d-justen/hyrise into mvcc_benchmark

julianmenzler committed:
Change implementation for physical_delete_loop() & add some debug logging

d-justen committed:
Add execution failure check, fix invalidation counter

julianmenzler committed:
Merge branch 'mvcc_delete_plugin' into mvcc_benchmark

# Conflicts:
#	src/plugins/mvcc_delete_plugin.cpp

d-justen committed:
Unuse locks

d-justen committed:
resolve merge conflicts

julianmenzler committed:
Fix multithreading issue

	-   When created, "std::unique_lock" automatically locks the assigned mutex. Afterwards we called the lock function again, which caused an error.
	-  We can prevent the automatic locking behaviour though by using the following constructor:
		std::unique_lock<std::mutex> queue_lock(_mutex_queue, std::defer_lock_t());

julianmenzler committed:
Improve lock efficiency

julianmenzler committed:
Implement condition variable for physical delete thread

d-justen committed:
Fix table chunk count

d-justen committed:
Make invalidated row counter mutable

julianmenzler committed:
Improve code - structured binding declaration

d-justen committed:
Dirty fix for segmentation fault

d-justen committed:
Merge branch 'mvcc_benchmark' of github.com:d-justen/hyrise into mvcc_benchmark

julianmenzler committed:
Implement PausableLoopThread to replace custom condition-variable approach

julianmenzler committed:
Add MvccDeletePlugin to CMakeLists.txt of playground & delete "dummy_mvcc_delete.cpp/hpp"

julianmenzler committed:
Fix linker issue in CMakeLists.txt for targets hyriseTest & hyriseSystemTest

julianmenzler committed:
Remove diry-fix for segmentation-fault (since it seems to be solved)

engelfa committed:
Fix linker error of pausable_loop_thread

julianmenzler committed:
Close #8: Update TableStatistics within delete_chunk()

julianmenzler committed:
Close #1: Replace FRIEND_TEST macros with regular "friend class" expression

julianmenzler committed:
Close #10: Merge Google Test Suites of MvccDeletePlugin

julianmenzler committed:
Format code

julianmenzler committed:
Make TransactionManager (_active_snapshot_commit_ids) thread-safe & add some comments (#18)

julianmenzler committed:
Add comment and small fixes & Delete dispensable timer.hpp

engelfa committed:
Close #5: Use get_table and delete get_referencing_table.

engelfa committed:
Add mutex header to transaction_manager.hpp.

engelfa committed:
Close #16: Describe input and output of script.

julianmenzler committed:
Fix TransactionManager DebugAssert

d-justen committed:
Fix #12 and #13

d-justen committed:
Delete old benchmarks #14 #15 #17

d-justen committed:
Test Update concurrently #11

julianmenzler committed:
Small enhancements regarding GetTable

julianmenzler commented on src/lib/operators/delete.hpp:
```diff
@@ -31,7 +32,7 @@ class Delete : public AbstractReadWriteOperator {
  private:
   TransactionID _transaction_id;
-
+  std::map<ChunkID, uint32_t> _num_rows_deleted_per_chunk;
```
I can't see any usage of this attribute. What its purpose?

engelfa commented on src/plugins/mvcc_delete_plugin.cpp:
```diff
@@ -0,0 +1,178 @@
+      _idle_delay_logical_delete(std::chrono::milliseconds(1000)),
+      _idle_delay_physical_delete(std::chrono::milliseconds(1000)) {}
+
+const std::string MvccDeletePlugin::description() const { return "This is the Hyrise TestPlugin"; }
```
```suggestion
const std::string MvccDeletePlugin::description() const { return "Physical MVCC delete plugin"; }
```

d-justen committed:
uncomment benchmark

d-justen commented on src/bin/CMakeLists.txt:
```diff
@@ -27,16 +27,18 @@ target_include_directories(
     ${CMAKE_DL_LIBS}
 )
 
+add_definitions(-DTEST_PLUGIN_DIR="${CMAKE_BINARY_DIR}/lib/")
```
It's the test plugin directory and needed for our benchmark to find the MvccDeletePlugin

julianmenzler replied:
ok, good to know 👍

d-justen commented on src/lib/operators/insert.cpp:
```diff
@@ -180,7 +180,7 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
 
     // while target chunk is not full
-    while (target_start_index != target_chunk->size()) {
+    while (target_start_index != target_chunk->size() && source_chunk_id < input_table_left()->chunk_count()) {
```
This wasn't supposed to be pushed. It was added because of our segmentation fault problem, but didn't solve the problem.

d-justen committed:
Merge branch 'mvcc_benchmark' of github.com:d-justen/hyrise into mvcc_benchmark

d-justen committed:
delete unnecessery condition

d-justen committed:
Add invalid row incrementation in chunks

d-justen committed:
Fix invalid row counting

julianmenzler committed:
Implement chunk metric invalid_row_count into MvccDeletePlugin & Delete operator (#21, #24)

commit db229ad130411c84a7a54b346c3563d2d008abc2
Author: Julian Menzler <julianmenzler@gmail.com>
Date:   Mon Feb 4 00:06:29 2019 +0100

    Remove logging and comments

commit 368562799c2ad84566b09e59b1a5ca8af36713cb
Author: Julian Menzler <julianmenzler@gmail.com>
Date:   Mon Feb 4 00:04:54 2019 +0100

    Change logical_delete_loop to only check chunks prior to the last one.

commit 6516bf18608cde8eb6f993449d24b38360542cd2
Author: Julian Menzler <julianmenzler@gmail.com>
Date:   Sun Feb 3 23:35:28 2019 +0100

    Remove wrong declaration in MvccDeletePlugin header

commit d12f174f2e80347fe2d889d3024d388ef813dcf6
Author: Julian Menzler <julianmenzler@gmail.com>
Date:   Sun Feb 3 23:30:42 2019 +0100

    Fix call of increase_invalidate_row_count() in Delete operator

commit 876350d7df35d1b7e0c988a311d60af9acb11ce9
Author: Julian Menzler <julianmenzler@gmail.com>
Date:   Sun Feb 3 23:06:15 2019 +0100

    Revert "Fix DebugAssert in _delete_chunk_logically"

    This reverts commit a9045cc89b842d400a22b1559da9432b0d3f8a0e.

commit a9045cc89b842d400a22b1559da9432b0d3f8a0e
Author: Julian Menzler <julianmenzler@gmail.com>
Date:   Sun Feb 3 22:25:12 2019 +0100

    Fix DebugAssert in _delete_chunk_logically

commit cc93447fbb5562c53a5367b1a470cad6c0d040a0
Author: Julian Menzler <julianmenzler@gmail.com>
Date:   Sun Feb 3 22:24:21 2019 +0100

    Use chunk invalid_row_count in MvccDeletePlugin

commit 2147e4da8e826ed3c09991dbadb7642f68cc6e7c
Author: Julian Menzler <julianmenzler@gmail.com>
Date:   Sun Feb 3 22:22:14 2019 +0100

    Enhance Delete operator to increase invalid_row_count of chunks

julianmenzler committed:
Format code & fix code style

engelfa committed:
Close #19: Add test case for logical delete.

julianmenzler commented on src/lib/storage/chunk.hpp:
```diff
@@ -162,6 +176,8 @@ class Chunk : private Noncopyable {
   pmr_vector<std::shared_ptr<BaseIndex>> _indices;
   std::shared_ptr<ChunkStatistics> _statistics;
   bool _is_mutable = true;
+  mutable uint64_t _invalid_row_count = 0;
```
done

engelfa committed:
Remove commented code in test.

julianmenzler committed:
Streamline code and naming

julianmenzler committed:
Streamline code of TransactionManager and use std::find & std::min_element

julianmenzler committed:
Add DebugAssert in class table, delete_chunk()

julianmenzler committed:
Improve DebugAsserts and naming in table.cpp & get_table.cpp

julianmenzler committed:
Implement test for GetTable: ExcludePhysicallyDeletedChunks (#26)

d-justen committed:
Resolve merge conflicts

d-justen committed:
Resolve merge conflicts pt. 2
